### ISSUE TITLE
[BUG] Inconsistent Nonce

### ISSUE THREAD

**Issue by yourbuddyconner** [CodaProtocol/coda#4935]
This bug is rearing its head again, seeing a veritable truckload of these logs in the `bugspray` network and no transactions are making it into blocks: 

```
{
 insertId: "qg0s93o03f79eg537"  
 jsonPayload: {
  level: "Debug"   
  message: "Decreasing trust for peer $peer_id due to Trust_system.Actions.Sent_useless_gossip (rejecting $cmd because of $reason). New trust is 0.000109."   
  metadata: {
   cmd: {
    payload: {
     body: [
      0: "Payment"       
      1: {
       amount: "1500000"        
       receiver_pk: "4vsRCVoPrFJCdX5XhEhE2HYC9o3WST6PLgUFTZjq2wfY6JAWwhdxxx1h6ENt4VeMLhk7ogjnqrqTsn71tb5dzPTnxU7vQ8mb6ptmdEEvNi5KkdxJWZFABgR3iCZck2bSK9WSWAZkndmyCGRy"        
       source_pk: "4vsRCVQeLHQPQuMrYkCHjke8JaG4rcCkVqYKsRJ4pszBgspDZVR3cDEsriUery5ek9wTtb7JFk9CVzuuyNBbJBmtNLVcgT21RBYt5k9kurMmELJ1bEKhiUB9i1vaxhNvW3X9aSrQtWYw1jQX"        
       token_id: "1"        
      }
     ]
     common: {
      fee: "0.06"       
      fee_payer_pk: "4vsRCVQeLHQPQuMrYkCHjke8JaG4rcCkVqYKsRJ4pszBgspDZVR3cDEsriUery5ek9wTtb7JFk9CVzuuyNBbJBmtNLVcgT21RBYt5k9kurMmELJ1bEKhiUB9i1vaxhNvW3X9aSrQtWYw1jQX"       
      fee_token: "1"       
      memo: "E4Yd67s51QN9DZVDy8JKPEoNGykMsYQ5KRiKpZHiLZTjA8dB9SnFT"       
      nonce: "342"       
      valid_until: "4294967295"       
     }
    }
    signature: "22Yqx7VderZhCbkRJrzPoi354LQ35QxrrWeHvMGQ1uGDWXAWeHNVKtRP6epipbPzTbKthKWfiLcqV3TEFWECNWSuan6vum5GTQ3mVexWC8ZWAFyS57QUsWCryZdxFFz2i4ywF5M9WXKskT2xQyppwn4n4m6BySWRSi6ecAgsNhqe9RgPmCQ1gGwXdDPZ3cVyAePxXK3FRdqxRZf7ADD3YKzffTgRz8PcMxFki9rr8F38be34F8bvBfERLrnvtAjBiogsRA3F8qs8ojC8aW"     
    signer: "4vsRCVQeLHQPQuMrYkCHjke8JaG4rcCkVqYKsRJ4pszBgspDZVR3cDEsriUery5ek9wTtb7JFk9CVzuuyNBbJBmtNLVcgT21RBYt5k9kurMmELJ1bEKhiUB9i1vaxhNvW3X9aSrQtWYw1jQX"     
   }
   host: "35.196.213.32"    
   peer_id: "12D3KooWKsbSpM2cPD22d4GremU6APHHh8HxuCj6M8ZxAyydK4AL"    
   pid: 10    
   port: 10001    
   reason: "invalid nonce"    
  }
  source: {
   location: "File "src/lib/trust_system/peer_trust.ml", line 135, characters 56-63"    
   module: "Trust_system__Peer_trust"    
  }
  timestamp: "2020-05-23 16:59:07.071499Z"   
 }
 labels: {
  k8s-pod/app: "whale-block-producer-1"   
  k8s-pod/class: "whale"   
  k8s-pod/pod-template-hash: "6bbd5c87fd"   
  k8s-pod/role: "block-producer"   
  k8s-pod/testnet: "bugspray"   
  k8s-pod/version: "0.0.12-beta-feature-bump-genesis-timestamp-4158ad2"   
 }
 logName: "projects/o1labs-192920/logs/stdout"  
 receiveTimestamp: "2020-05-23T16:59:13.411477159Z"  
 resource: {
  labels: {
   cluster_name: "coda-infra-east"    
   container_name: "coda"    
   location: "us-east1"    
   namespace_name: "bugspray"    
   pod_name: "whale-block-producer-1-6bbd5c87fd-p596s"    
   project_id: "o1labs-192920"    
  }
  type: "k8s_container"   
 }
 severity: "INFO"  
 timestamp: "2020-05-23T16:59:07.222639472Z"  
}
```

When examining the sender account via the graphql-proxy, the nonce is `11`, indicating some major level of inconsistency in the mempool. 

```
{
  "data": {
    "version": "4158ad2061345c594b9237efdb1f32118406c1ab",
    "syncStatus": "SYNCED",
    "account": {
      "nonce": "11",
      "publicKey": "4vsRCVQeLHQPQuMrYkCHjke8JaG4rcCkVqYKsRJ4pszBgspDZVR3cDEsriUery5ek9wTtb7JFk9CVzuuyNBbJBmtNLVcgT21RBYt5k9kurMmELJ1bEKhiUB9i1vaxhNvW3X9aSrQtWYw1jQX"
    }
  }
}
```

Mentioned in: 
#4066

**Comment by deepthiskumar** [comment]
These were initially caused because transaction fees were the same in all the transactions produced by the nodes. So transactions that get discarded were not uniform across. We added random transaction fees to mitigate this. However, we still invalid nonce errors. After looking at various instances of the invalid nonces error, the common pattern is that not all the nodes received all the transactions. To confirm this further, I looked for the trust system log that is logged when a transaction is accepted into the pool since this is unique per transaction as opposed to the "received transaction" log that gets logged every time a peer sends the txn. Looked at 30 transactions from different source pks and with nonce = 0. 
With nonce = 0, there is no reason why a transaction wouldn't be accepted given these are the first set of transactions sent and that the pool is still not full yet. And so, if there is no log that says this is accepted then the node must have not received it. Also, this considers the rebroadcasting scenario as well. Currently the rebroadcast window is 50mins from the time the transaction was included in the source node's pool. A transaction is rebroadcasted 5 times within this window.
Here's the file that, for each transaction, has the number of nodes that accepted the transaction once or twice.
[zero-nonce-txns-accepted.txt](https://github.com/CodaProtocol/coda/files/4737998/zero-nonce-txns-accepted.txt)
The number of nodes that receive a transaction are 164/165 in a 205-node network. Since some nodes don't receive certain transactions, their view of transactions with lowest fee is different among nodes and therefore, transaction evictions due to max size also differ causing invalid nonce errors.
In the file, there are some transactions that are accepted twice. This is because:
1. the nodes that produced the transactions evicted their own transactions because of low fees and generated new set of transactions. The other nodes that also evicted in a similar fashion received the new set of transactions and accepted them again (in the case the transaction pool is full, it gets accepted if there are other lower fee transactions that can be evicted). The nodes that only accept once either haven't evicted the transaction from the pool the first time or they did not receive it the second time.
2. The node crash and loses all its transactions in the pool and starts generating new set of transactions from the nonce in the account. Such transactions mostly get rejected due to insufficient replace fee
